cpp/src/arrow/type_traits.cc

```diff
@@ -0,0 +1,96 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/type_traits.h"
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+int RequiredValueAlignmentForBuffer(Type::type type_id, int buffer_index) {
+  if (buffer_index != 1) {
+    // If the buffer index is 0 then either:
+    //  * The array type has no buffers, thus this shouldn't be called anyways
+    //  * The array has a validity buffer at 0, no alignment needed
+    //  * The array is a union array and has a types buffer at 0, no alignment needed
+    // If the buffer index is > 1 then, in all current cases, it represents binary
+    //  data and no alignment is needed
```

Comment:
Hmm, I took a look again and it's more complicated. Unions don't have a null bitmap, but in Arrow C++ we still reserve an empty slot in buffer 0 for the non-existent union null bitmap. Therefore, the dense union offsets buffer is number 2, not 1, and needs to be word-aligned...

Reply:
I was not aware of that.  I'll add a test case for this.

Reply:
I've fixed dense unions and test these independently now.